src/export/CodeSystemExporter.ts

```diff
@@ -1,15 +1,26 @@
 import { FSHTank } from '../import/FSHTank';
-import { CodeSystem, CodeSystemConcept } from '../fhirtypes';
+import { CodeSystem, CodeSystemConcept, validateFHIRId } from '../fhirtypes';
 import { FshCodeSystem } from '../fshtypes';
 import { logger } from '../utils/FSHLogger';
 import { Package } from '.';
+import { validateFHIRName } from '../fhirtypes/common';
 
 export class CodeSystemExporter {
   constructor(private readonly tank: FSHTank, private readonly pkg: Package) {}
 
   private setMetadata(codeSystem: CodeSystem, fshDefinition: FshCodeSystem): void {
     codeSystem.name = fshDefinition.name;
     codeSystem.id = fshDefinition.id;
+    try {
+      validateFHIRName(codeSystem.name);
+    } catch (ex) {
+      logger.error(ex.message, fshDefinition.sourceInfo);
+    }
+    try {
+      validateFHIRId(codeSystem.id);
+    } catch (ex) {
+      logger.error(ex.message, fshDefinition.sourceInfo);
+    }
```

Comment:
These take up a lot of lines of code to do validation.  Did you consider other alternatives?

For example, you could make `validateFHIRName` and `validateFHIRId` take an optional `SourceInfo` as well and return back the name/id that was passed in.  Then you can log the message in the validation function itself -- and that would also allow you to use it like this:
```ts
codeSystem.name = validateFHIRName(fshDefinition.name, fshDefinition.sourceInfo);
codeSystem.id = validateFHIRId(fshDefinition.id, fshDefinition.sourceInfo);
```

That would make a lot less noise and also allow you to remove those new Error classes (they wouldn't really be needed anymore).  I like having a place to put the refs to the docs, but we're not actually leveraging those right now, so you could just put them in as comments in the validate function for now.  Eventually we can figure out how to integrate them into the logging framework.

Whaddya think?  Like it or is it too weird?

Reply:
Hmm, how about this: there could be functions for `setName` and `setId` on the various FHIR classes. In those functions, the value would be assigned, and they would call the `validateFHIRName` and `validateFHIRId` functions. The validate functions would be changed to return a boolean result instead of throwing, so that would let me get rid of the two error classes as you mentioned. With this approach, the actual exporters would just call the set functions. This would also have the advantage that I know what FHIR resource type I'm working with, which lets me write more specific error messages or link to more specific documentation pages. The downside of this approach is that I'm throwing yet another function call on to the stack, and that the FHIR types would have nearly-identical set functions.

Maybe this could be a situation where it finally makes sense to make an abstract supertype for the various FHIR resources that get exported? Ugh, but Instances don't have names, which means that one way or another, things get a little ugly.

Reply:
Would the `setName`/`setId` functions also take in a `SourceInfo` so they can report the lines?  I don't think we want to lose that functionality.  If so, then I'm good with this approach as well!

Reply:
Upon further consideration, this seems like a reasonable time for me to try using mixins. The next commit will contain my attempt to do so.

Reply:
Woah.  Looking forward to it!